acasclient/lsthing.py

```diff
@@ -971,23 +961,43 @@ def _validate_params(self, code, code_type, code_kind, code_origin, client):
         :type code_kind: str
         :param code_origin: Origin of DDictValue referenced, typically 'ACAS DDict'
         :type code_origin: str
-        :param client: Authenticated acasclient.client instance to look up current DDictValues
-        :type client: acasclient.client
-        :return: Error message, or None if valid
-        :rtype: str | None
+        :param ddict: DDict object being referenced. Can be used instead of code_type + code_kind + code_origin
+        :type ddict: DDict
         """
-        if client is None:
-            return
-        valid_val_dicts = client.get_ddict_values_by_type_and_kind(
-            code_type, code_kind)
-        if valid_val_dicts == []:
-            return (f"Invalid 'code_type':'{code_type}' or "
-                    f"'code_kind':'{code_kind}' provided")
-        if any([code == val_dict['code'] for val_dict in valid_val_dicts]):
-            return
-
-        return (f"Invalid 'code':'{code}' provided for the given "
-                f"'code_type':'{code_type}' and 'code_kind':'{code_kind}'")
+        self.code = code
+        self.ddict = ddict
+        if ddict:
+            if isinstance(ddict, DDict):
+                self.code_type = ddict.code_type
+                self.code_kind = ddict.code_kind
+                self.code_origin = ddict.code_origin
+            else:
+                raise ValueError('ddict must be an instance of acasclient.ddict.DDict')
+        else:
+            self.code_type = code_type
+            self.code_kind = code_kind
+            self.code_origin = code_origin
+            # Auto-recognize some code origin:
+            if self.code_origin.upper() == ACAS_DDICT:
+                self.ddict = ACASDDict(code_type, code_kind)
+            elif self.code_origin.upper() == ACAS_LSTHING:
+                self.ddict = ACASLsThingDDict(code_type, code_kind)
+
+    def __hash__(self):
+        return hash(f'{self.code}-{self.code_type}-{self.code_kind}-'
+                    f'{self.code_origin}')
+
+    @validation_result
+    def validate(self):
+        """Validate that this CodeValue conforms to the stored list of possible DDictValues in self.ddict
+
+        :return: ValidationResult indicating True / valid if code is valid, False / invalid otherwise
+        :rtype: ValidationResult
+        """
+        if self.ddict:
+            return self.ddict.check_value(self.code)
+        else:
+            return True
```

Comment:
Optional: We don't necessarily need it. If the function return value is `None` it will automatically be type cast to `ValidationResult(True, [])`.